enhancements/machine-config/on-cluster-builds.md

```diff
@@ -0,0 +1,949 @@
+  workflows, etc. If those features are required, it is recommended that
+  a cluster administrator use an off-cluster image build mechanism
+  instead.
+
+- There are many interesting conversations happening around integrating
+  image-based RHEL into OCP. A lot of these conversations have to do with
+  defining or redefining what promises RHCOS makes with regard to OS
+  immutability. While on-cluster builds must be aware of image-based RHEL
+  with that eventuality in mind, figuring out future OS immutability
+  capabilities is out-of-scope for this enhancement.
+
+- There will be situations where the MCO will need to work around feature gaps
+  in bootc. The MCO would not be alone since RHEL For Edge and RHIVOS are likely
+  to face similar situations and friction. While these concerns exist,
+  addressing them within the context of RHEL For Edge and RHIVOS is out-of-scope
+  for this enhancement.
+
+## Proposal
+
+### Workflow Description
+
+1.  The cluster administrator realizes that they need to add custom
+    content to the OS on each of their cluster nodes.
+
+2.  The cluster administrator edits the MachineConfigPool, adding a
+    Dockerfile, as well as the one-time setup of what image registry to push the
+    final image to, the credentials needed to do so, etc.
+
+3.  The BuildController detects that these details are provided and
+    begins a build.
+
+4.  The built image is pushed to the cluster administrators’ image
+    registry of choice.
+
+5.  NodeController updates the `desiredImage` annotation on each node within its
+    pool according to its rules such as maximum unavailable nodes.
+
+6.  The MachineConfigDaemon drains the target node, applies the new OS
+    content to the node, and reboots the node upon completion.
+
+#### Variation and form factor considerations
+
+##### Single-Node OpenShift (SNO)
+
+Single-Node OpenShift (SNO) should not have any difficulty running the
+on-cluster build process as-is. However, there may be some efficiency
+improvements that can be made. For example, the final image could be written
+directly to the nodes’ filesystem before the handoff to rpm-ostree / bootc to
+avoid a registry roundtrip.
+
+##### Hypershift
+
+Hypershift on the other hand, would require distinct configurations for
+both the host cluster and the guest cluster. Additionally, some
+adaptations would be required to Hypershift in order to traverse the
+host \<-\> guest cluster barrier. (need to page Hypershift MCO process
+back in to have a more complete opinion around this).
+
+##### Multi-Arch Clusters
+
+Multiarch clusters have special considerations in terms of needing to
+orchestrate more than one build for each target architecture as well as
+manifestlisting the final image. This consideration also extends to the
+image registry the final image will be pushed to. To my understanding,
+the default on-cluster image registry does not support manifestlisted
```

Comment:
the internal image registry does support pushing and pulling manifestlists now . Imagestream imports was not working with manifestlists and that has been introduced since 4.13

Reply:
That's great to know! I'll adjust the enhancement accordingly. I'll also do some separate experimentation around this once we begin looking at multiarch support.